docs/syscall_descriptions_syntax.md

```diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
 define MY_PATH_MAX	PATH_MAX + 2
 ```
 
+## Conditional fields
+
+### In structures
+
+In syzlang, it's possible to specify a condition for every struct field that
+determines whether the field should be included or omitted:
+
+```
+header_fields {
+  magic       const[0xabcd, int16]
+  haveInteger int8
+} [packed]
+
+packet {
+  header  header_fields
+  integer int64  (if[value[header:haveField] == 0x1])
+  body    array[int8]
+} [packed]
+
+some_call(a ptr[in, packet])
+```
+
+In this example, the `packet` structure will include the field `integer` only
+if `header.haveInteger == 1`. That corresponds to e.g. the following program:
+```
+some_call(&AUTO={{AUTO, 0x1}, @value=0xabcd, []})
+```
+
+If `header.haveInteger` is not `1`, syzkaller will just pretend that the field
+`integer` does not exist -- much like it's done for `optional[T]`.
+```
+some_call(&AUTO={{AUTO, 0x0}, @void, []})
+```
+
+A structure that has conditional fields is assumed to be `[varlen]` and.  If
+a field of a `varlen` type or a conditional field appears in the middle of a
+structure, the structure must be `[packed].`
+
+Conditions on bitfields are prohibited:
+```
+struct {
+  f0 int
+  f1 int:3 (if[value[f0] == 0x1])  # It will not compile.
+}
+```
+
+But you may reference bitfields in your conditions:
+```
+struct {
+  f0 int:1
+  f1 int:7
+  f2 int   (if[value[f0] == value[f1]])
+} [packed]
+```
+
+### In unions
+
+Let's consider the following example.
+
+```
+struct {
+  type int
+  body alternatives
+}
+
+alternatives [
+  int     int64 (if[value[struct:type] == 0x1])
+  arr     array[int64, 5] (if[value[struct:type] == 0x1])
+  default int32
+] [varlen]
+
+some_call(a ptr[in, struct])
+```
+
+In this case, the specific union field will be selected depending on the value
+of the `type` field. For example, if `type` is `0x1`, then it's `int`:
+```
+some_call(&AUTO={0x1, @int=0x1})
+```
+
+If no condition matches, syzkaller would pick the last filed that **must always
+have no condition**.
+
+```
+some_call(&AUTO={0x0, @default=0xabcd})
+```
+
+The following definition would fail to compile:
+
+```
+alternatives [
+  int int64 (if[value[struct:type] == 0x1])
+  arr array[int64, 5] (if[value[struct:type] == 0x1])
+] [varlen]
+```
+
+Syzkaller will always obey these restrictions. Specifically, it will select the
+first union field whose condition is satisfied.
```

Comment:
Hm, wouldn't it make more sense to randomly select one field out of the ones that satisfy the condition if there are many of them?

Reply:
This semantics was necessary to implement conditional struct fields on top of conditional union fields.

If we go with `take random union field that matches`, then the field without condition may always be taken, so the current transformation:

`field type (if[A])` =>
```
[
value type (if[A])
void void
]
```

no longer fully works. If we transform to (*)
```
[
value type (if[A])
void void (if[(A) == 0])
]
```

then we need to support unions with conditions on all fields. Then:
* We have trouble designating the default union field. 
* (Specifically) We must be able to prove that at least one of the conditions is always satisfiable. It can be tricky.

So the current semantics was just simpler.

Theoretically we could do the following though:
* For user-defined unions, continue to require the last option to have no if condition.
* Transform structs with conditional fields as in (*), i.e. so that all union options do have conditions. We know that one of the conditions is always true, so it will work.

Reply:
I don't get what you meant in the part about transformations, but if a union field is now selected randomly - that's perfect!